### ISSUE TITLE
Column ordering matters when converting to XML

### ISSUE THREAD

**Issue by pralitp** [JGCRI/gcamdata#369]
We have thus far been flexible with regards to column ordering, allowing the order to differ from the old data system so long as the content was the same.

Unfortunately converting tables to XML does currently assume some column ordering since the Model Interface "header" maps columns based on column positions ignoring what the actual column names in the table being converted are.

I could see two possible solutions to this.  First is what was essentially done in the old data system: maintain a "level2_data_names" which contains a database of column name ordering as expected in the Model Interface headers as written.  In the old data system we used to re-order columns based off of these as the last step before doing CSV to XML conversions.

An alternative could be to change the behavior of the CSV to XML tool so that it does in fact check the column names in the data to be converted and permute the columns to match columns in the header file as written.  This implies that the column names in our Level 2 tibbles must match the column names as written in the header which may not always be the case at present.

**Comment by pralitp** [comment]
I could imagine @pkyle and @rplzzz may have ideas on this issue.

**Comment by pkyle** [comment]
I don't see any way that the model interface could always correctly guess the column ordering from the strings we're using as column names, at least not unless we make the column names much more verbose. For example, where we currently name a column "year", we'd need to name it "technology/+{year}period", or "subsector/+{year}share-weight", or whatever the correct string is for the given header/column.
What I'd think of doing if we're looking to automate this is to programmatically translate the headers in ModelInterface_headers.txt into column names. I'd imagine we're almost always consistent there (e.g., the column name for "world/+{name}region" is always "region"). One place where I know we'd have exceptions is the CalOutput and CalInput tables, where subsector and technology share-weights are read using different strings than what is in the header and XML file ("subs.share.weight" and "tech.share.weight", because we don't want two columns with the same name). Of course doing this programatically might take more time than just keying them in manually...

**Comment by rplzzz** [comment]
Keeping a database of expected column orderings seems like the lest bad solution in this case.  It should be easy to supply a utility subroutine to reorder the columns according to what's in the DB.

In the long run my concern with all of this is that anybody attempting to add a feature to GCAM is going to have a very hard time figuring out how to get the XML conversion to work.  For purposes of getting the conversion working we can just let it go, but at some point I think we are going to want to do some thinking about how we specify our input database schema.

**Comment by pralitp** [comment]
Yea, I didn't fully appreciate all of the cases where trying to match column names would be difficult as Page points out.

So I agree with Robert.  Keeping the Level 2 data names database is the least bad solution.  Given we already have this from the old data system getting it up and running won't be difficult.  Here is what I am thinking:  Write some script to translate the current definitions that look like:
```
names_DepRsrc <- c( "region", "depresource", "output.unit", "price.unit", "market" )
```
To:
```
list(
  "DepRsrc" = c( "region", "depresource", "output.unit", "price.unit", "market" )
)
```

That way we can lookup the column name ordering using the same "tag" as the header in "ModelInterface_headers.txt".  We can then just do the column ordering translation by changing `add_xml_data` to do it:
```
add_xml_data <- function(dot, data, header, column_order_lookup = header) {
  # Users can skip column reordering by passing NULL as the argument value.
  if(!isNULL(column_order_lookup)) {
    data <- data[, LEVEL2_DATA_NAMES[[column_order_lookup]] ]
  }
  curr_table <- list(data = data, header = header)
  dot$data_tables[[length(dot$data_tables)+1]] <- curr_table

  dot
}
```

Thoughts?

**Comment by pralitp** [comment]
With regards to maintaining into the future maybe we should open a separate issue?  I think that is just a question of documentation since setting up the conversion isn't really that complicated.  It basically boils down to setting up a mapping from columns in a table to XML tag names used in GCAM including parent/child relationship.

**Comment by pkyle** [comment]
This all sounds good to me--wrt the issue about constructing the "header" list, note that a lot of the level2 data tables don't have their own R object in level2_data_names, and conversely, a lot of the objects in level2_data_names.R are only used in the processing code when concatenated with other strings (generally keyed manually), for column selection and ordering. These objects in level2_data_names.R don't need to be in the global header list. For example, tables using the header TechShrwt generally have their columns selected and ordered as follows:
data_table <- data_table[ c( names_TechYr, "share-weight" ) ]
In other words, there's no names_TechShrwt object, and the names_TechYr object that does exist doesn't actually correspond to a header that is used by any data tables. There are a bunch of examples of this. Just FYI for writing a script to make the header list. And if it's gotta be done manually I can do that.
And wrt the broader aspects of the method, as long as we have documentation demonstrating how to add headers to the list, I don't see this as being a problem.

**Comment by rplzzz** [comment]
@pralitp Definitely a matter for the future.  I just wanted to get it on your radar.

**Comment by pralitp** [comment]
Ok, so I'd dropped the ball on this one for a while.  So I followed the approach I suggested earlier:

1. Wrote a script to convert level 2 data names from the format in the old system `names_XYZ <- c("X", "Y", "Z")` to `LEVEL2_DATA_NAMES[["XYZ"]] <- c("X", "Y", "Z")` (lvl2_mapped)
2. As @pkyle noted there were a lot of missing headers so I wrote a script to search all level 2 CSV files and generate the names from the column names seen there (lvl2_mapped_csv).  Although there were still some missing or had duplicative ways they were written to CSV (errors) which I manually fixed up.

I've pushed up my branch `xml-column-ordering`  if all looks ok I'll clean up the ordering of the level 2 headers then open a pull request.

[errors.txt](https://github.com/JGCRI/gcamdata/files/1065166/errors.txt)
[lvl2_mapped_csv.txt](https://github.com/JGCRI/gcamdata/files/1065168/lvl2_mapped_csv.txt)
[lvl2_mapped.txt](https://github.com/JGCRI/gcamdata/files/1065167/lvl2_mapped.txt)

**Comment by pkyle** [comment]
Looks great! Just remembered, we'll need to do something to make sure the "node_rename" option is still available, and is used on the appropriate tables in the land allocator. Or there might be a different way of dropping the numbers on the LandNode values.

**Comment by pralitp** [comment]
Yup, good call.  We still have to do deal with the numbers on the LandNode values.  I'll just add a new function that can be used in the "create XML pipeline" to do that.
